consensus/src/lib.rs

```diff
@@ -1,27 +1,1101 @@
+            }
+        }
+
+        // Connect the new chain
+        for block_index in new_chain {
+            self.connect_tip(&block_index, &best_block_id)?;
+        }
+        Ok(())
+    }
+
+    fn store_cached_inputs(&mut self, cached_inputs: &CachedInputs) -> Result<(), BlockError> {
+        for (tx_id, tx_index) in cached_inputs.iter() {
+            self.db_tx.set_mainchain_tx_index(tx_id, tx_index)?;
+        }
+        Ok(())
+    }
+
+    fn calculate_indices(
+        &self,
+        block: &Block,
+        tx: &Transaction,
+    ) -> Result<TxMainChainIndex, BlockError> {
+        let enc_block = block.encode();
+        let enc_tx = tx.encode();
+        let offset_tx = enc_block
+            .windows(enc_tx.len())
+            .enumerate()
+            .find_map(|(i, d)| (d == enc_tx).then(|| i))
+            .ok_or(BlockError::Unknown)? as u32;
+
+        let tx_position =
+            TxMainChainPosition::new(&block.get_id().get(), offset_tx as u32, enc_tx.len() as u32);
+
+        assert_eq!(
+            &self.db_tx.get_mainchain_tx_by_position(&tx_position).ok().flatten().expect("Database corrupted! "),
+            tx
+        );
+
+        TxMainChainIndex::new(
+            SpendablePosition::from(tx_position),
+            tx.get_outputs().len() as u32,
+        )
+        .map_err(BlockError::from)
+    }
+
+    fn connect_transactions(&mut self, block: &Block) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in block.transactions().iter() {
+            for input in tx.get_inputs() {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
+                    Some(tx_index) => tx_index.clone(),
+                    None => {
+                        let tx_index = self.calculate_indices(block, tx)?;
+                        cached_inputs.insert(tx.get_id(), tx_index.clone());
+                        tx_index
+                    }
+                };
+                if input_index >= tx_index.get_output_count() {
+                    return Err(BlockError::Unknown);
+                }
+                // Set each input as spent
+                tx_index
+                    .spend(input_index, Spender::from(tx.get_id()))
+                    .map_err(BlockError::from)?;
+                cached_inputs.insert(tx.get_id(), tx_index.clone());
+            }
+        }
+        self.store_cached_inputs(&cached_inputs)?;
+        Ok(())
+    }
+
+    fn disconnect_transactions(&mut self, transactions: &[Transaction]) -> Result<(), BlockError> {
+        let mut cached_inputs = CachedInputs::new();
+        for tx in transactions.iter().rev() {
+            let inputs = tx.get_inputs();
+            for input in inputs {
+                let input_index = input.get_outpoint().get_output_index();
+
+                let mut tx_index = match cached_inputs.get(&tx.get_id()) {
```

Comment:
Mirror image of my comment on `connect_transactions` :)

Reply:
Fixed!